services/QuillLMS/engines/evidence/app/services/evidence/research/gen_ai/confusion_matrix_builder.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module Evidence
+  module Research
+    module GenAI
+      class ConfusionMatrixBuilder < ApplicationService
+        attr_reader :llm_examples
+
+        OPTIMAL = 0
+        SUBOPTIMAL = 1
+
+        def initialize(llm_examples)
+          @llm_examples = llm_examples
+        end
+
+        # results is a 2x2 matrix where
+        # rows represent the ground truth (optimal or suboptimal),
+        # columns represent the model prediction (optimal or suboptimal),
+        def run
+          [[0, 0], [0, 0]].tap do |matrix|
+            llm_examples.each do |llm_example|
```

Comment:
There are a few ways to make this shorter, but if you want this row/column setup you could do this in the loop instead:
```ruby
row = llm_example.optimal? ? OPTIMAL : SUBOPTIMAL
column = llm_example.test_optimal? ? OPTIMAL : SUBOPTIMAL

matrix[row][column] += 1

Reply:
Thanks!  That's much cleaner.